File: crates/json-schema-transformer/src/emit/pydantic/jsi_validator.py

import math as _jsi_math
import sys as _jsi_sys
from urllib.parse import unquote as _jsi_unquote, urljoin as _jsi_urljoin

if _jsi_sys.getrecursionlimit() < 10000:
    _jsi_sys.setrecursionlimit(10000)

_JSI_UNDEF = object()
_JSI_SCHEME_RE = re.compile(r"^[A-Za-z][A-Za-z0-9+.\-]*:")


class _JsiLimit(Exception):
    pass


def _jsi_num(x):
    return isinstance(x, (int, float)) and not isinstance(x, bool)


def _jsi_equal(a, b):
    if isinstance(a, bool) or isinstance(b, bool):
        return isinstance(a, bool) and isinstance(b, bool) and a == b
    if _jsi_num(a) and _jsi_num(b):
        return a == b
    if type(a) is not type(b):
        return False
    if isinstance(a, dict):
        return a.keys() == b.keys() and all(_jsi_equal(a[k], b[k]) for k in a)
    if isinstance(a, list):
        return len(a) == len(b) and all(_jsi_equal(x, y) for x, y in zip(a, b))
    return a == b


def _jsi_strip_frag(u):
    i = u.find("#")
    return u if i == -1 else u[:i]


def _jsi_resolve_uri(base, ref):
    if ref.startswith("#"):
        return _jsi_strip_frag(base) + ref
    if _JSI_SCHEME_RE.match(ref):
        return ref
    try:
        resolved = _jsi_urljoin(base, ref)
        return resolved if resolved else ref
    except Exception:
        return ref


def _jsi_compile_re(pat):
    try:
        return re.compile(pat)
    except Exception:
        return None


def _jsi_ptr_get(doc, ptr):
    if ptr == "":
        return doc
    cur = doc
    for raw in ptr.split("/")[1:]:
        try:
            seg = _jsi_unquote(raw)
        except Exception:
            seg = raw
        seg = seg.replace("~1", "/").replace("~0", "~")
        if isinstance(cur, list):
            try:
                i = int(seg)
            except (ValueError, TypeError):
                return _JSI_UNDEF
            if i < 0 or i >= len(cur):
                return _JSI_UNDEF
            cur = cur[i]
        elif isinstance(cur, dict):
            if seg not in cur:
                return _JSI_UNDEF
            cur = cur[seg]
        else:
            return _JSI_UNDEF
    return cur


def _jsi_validate(root_schema, remotes, instance):
    resources = {}
    anchors = {}
    dyn_anchors = {}
    base_of = {}
    skip_keys = {"const", "enum", "default", "examples"}
    name_maps = {"properties", "patternProperties", "$defs", "definitions", "dependentSchemas"}

    def index_schema(node, base):
        if isinstance(node, list):
            for n in node:
                index_schema(n, base)
            return
        if not isinstance(node, dict):
            return
        cur = base
        nid = node.get("$id")
        if isinstance(nid, str):
            cur = _jsi_strip_frag(_jsi_resolve_uri(base, nid))
            resources[cur] = node
        base_of[id(node)] = cur
        anc = node.get("$anchor")
        if isinstance(anc, str):
            anchors[cur + "#" + anc] = node
        dyn = node.get("$dynamicAnchor")
        if isinstance(dyn, str):
            dyn_anchors.setdefault(cur, {})[dyn] = node
            anchors[cur + "#" + dyn] = node
        for k, v in node.items():
            if k in skip_keys:
                continue
            if k in name_maps:
                if isinstance(v, dict):
                    for s in v.values():
                        index_schema(s, cur)
                continue
            index_schema(v, cur)

    root_base = "urn:jsi:root"
    if isinstance(root_schema, dict):
        rid = root_schema.get("$id")
        if isinstance(rid, str):
            root_base = _jsi_strip_frag(_jsi_resolve_uri("urn:jsi:root", rid))
    resources[root_base] = root_schema
    index_schema(root_schema, root_base)
    for uri, doc in remotes.items():
        u = _jsi_strip_frag(uri)
        if u not in resources:
            resources[u] = doc
        index_schema(doc, u)

    # Honor the dialect's $vocabulary: if the validation vocabulary is absent,
    # assertion keywords (type, const, minimum, required, ...) do not apply.
    assertions = True
    if isinstance(root_schema, dict):
        dialect = root_schema.get("$schema")
        if isinstance(dialect, str):
            meta = resources.get(_jsi_strip_frag(dialect))
            if isinstance(meta, dict):
                vocab = meta.get("$vocabulary")
                if isinstance(vocab, dict):
                    assertions = any(
                        ("/vocab/validation" in k) and v is not False
                        for k, v in vocab.items()
                    )

    def resolve_ref(ref, base):
        uri = _jsi_resolve_uri(base, ref)
        hi = uri.find("#")
        frag = uri[hi + 1:] if hi != -1 else ""
        doc_uri = _jsi_strip_frag(uri)
        if frag != "" and not frag.startswith("/"):
            target = anchors.get(doc_uri + "#" + frag)
            if target is None:
                return None
            return target, base_of.get(id(target), doc_uri)
        doc = resources.get(doc_uri, _JSI_UNDEF)
        if doc is _JSI_UNDEF:
            return None
        target = _jsi_ptr_get(doc, frag)
        if target is _JSI_UNDEF:
            return None
        tbase = base_of.get(id(target)) if isinstance(target, (dict, list)) else None
        return target, (tbase if tbase is not None else doc_uri)

    depth = [0]

    def new_ann():
        return {"props": set(), "prefix": 0, "all": False, "idxs": set()}

    def vs(schema, inst, base, dscope, ann):
        if isinstance(schema, bool):
            return schema
        if not isinstance(schema, dict):
            return True
        depth[0] += 1
        if depth[0] > 512:
            raise _JsiLimit()
        try:
            s = schema
            my_base = base_of.get(id(schema), base)
            scope = dscope
            if not scope or scope[-1] != my_base:
                scope = dscope + [my_base]

            local = new_ann()

            def merge(child):
                local["props"] |= child["props"]
                local["idxs"] |= child["idxs"]
                if child["prefix"] > local["prefix"]:
                    local["prefix"] = child["prefix"]
                local["all"] = local["all"] or child["all"]

            def sub(sch, i):
                a = new_ann()
                return vs(sch, i, my_base, scope, a), a

            # --- $ref / $dynamicRef (in-place applicators) ---
            ref = s.get("$ref")
            if isinstance(ref, str):
                r = resolve_ref(ref, my_base)
                if r is None:
                    return False
                a = new_ann()
                if not vs(r[0], inst, r[1], scope, a):
                    return False
                merge(a)

            dref = s.get("$dynamicRef")
            if isinstance(dref, str):
                r = resolve_ref(dref, my_base)
                if r is None:
                    return False
                target, tbase = r
                hi = dref.find("#")
                frag = dref[hi + 1:] if hi != -1 else ""
                is_plain = frag != "" and not frag.startswith("/")
                if is_plain and isinstance(target, dict) and target.get("$dynamicAnchor") == frag:
                    for res in scope:
                        m = dyn_anchors.get(res)
                        if m is not None and frag in m:
                            target = m[frag]
                            tbase = base_of.get(id(target), res)
                            break
                a = new_ann()
                if not vs(target, inst, tbase, scope, a):
                    return False
                merge(a)

            # --- type ---
            if assertions and "type" in s:
                tval = s["type"]
                types = tval if isinstance(tval, list) else [tval]
                ok = False
                for t in types:
                    if t == "null":
                        ok = inst is None
                    elif t == "boolean":
                        ok = isinstance(inst, bool)
                    elif t == "string":
                        ok = isinstance(inst, str)
                    elif t == "number":
                        ok = _jsi_num(inst)
                    elif t == "integer":
                        ok = _jsi_num(inst) and (isinstance(inst, int) or inst.is_integer())
                    elif t == "array":
                        ok = isinstance(inst, list)
                    elif t == "object":
                        ok = isinstance(inst, dict)
                    else:
                        ok = False
                    if ok:
                        break
                if not ok:
                    return False

            # --- const / enum ---
            if assertions and "const" in s and not _jsi_equal(inst, s["const"]):
                return False
            if assertions and isinstance(s.get("enum"), list) and not any(
                _jsi_equal(inst, e) for e in s["enum"]
            ):
                return False

            # --- numeric ---
            if assertions and _jsi_num(inst):
                mo = s.get("multipleOf")
                if _jsi_num(mo):
                    try:
                        q = inst / mo
                        if not _jsi_math.isfinite(q) or q != _jsi_math.floor(q):
                            return False
                    except (OverflowError, ZeroDivisionError):
                        return False
                mn = s.get("minimum")
                if _jsi_num(mn) and inst < mn:
                    return False
                mx = s.get("maximum")
                if _jsi_num(mx) and inst > mx:
                    return False
                emn = s.get("exclusiveMinimum")
                if _jsi_num(emn) and inst <= emn:
                    return False
                emx = s.get("exclusiveMaximum")
                if _jsi_num(emx) and inst >= emx:
                    return False

            # --- string ---
            if assertions and isinstance(inst, str):
                ml = s.get("minLength")
                if _jsi_num(ml) and len(inst) < ml:
                    return False
                mxl = s.get("maxLength")
                if _jsi_num(mxl) and len(inst) > mxl:
                    return False
                pat = s.get("pattern")
                if isinstance(pat, str):
                    rex = _jsi_compile_re(pat)
                    if rex is not None and rex.search(inst) is None:
                        return False

            # --- array ---
            if isinstance(inst, list):
                if assertions:
                    mi = s.get("minItems")
                    if _jsi_num(mi) and len(inst) < mi:
                        return False
                    mxi = s.get("maxItems")
                    if _jsi_num(mxi) and len(inst) > mxi:
                        return False
                    if s.get("uniqueItems") is True:
                        for i in range(len(inst)):
                            for j in range(i + 1, len(inst)):
                                if _jsi_equal(inst[i], inst[j]):
                                    return False
                prefix = s.get("prefixItems")
                if not isinstance(prefix, list):
                    prefix = []
                for i in range(min(len(prefix), len(inst))):
                    if not sub(prefix[i], inst[i])[0]:
                        return False
                if len(prefix) > 0:
                    local["prefix"] = max(local["prefix"], min(len(prefix), len(inst)))
                if "items" in s:
                    items = s["items"]
                    for i in range(len(prefix), len(inst)):
                        if not sub(items, inst[i])[0]:
                            return False
                    local["all"] = True
                if "contains" in s:
                    contains = s["contains"]
                    matched = [i for i in range(len(inst)) if sub(contains, inst[i])[0]]
                    minc = s.get("minContains")
                    minc = minc if _jsi_num(minc) else 1
                    maxc = s.get("maxContains")
                    if len(matched) < minc:
                        return False
                    if _jsi_num(maxc) and len(matched) > maxc:
                        return False
                    local["idxs"] |= set(matched)

            # --- object ---
            if isinstance(inst, dict):
                keys = list(inst.keys())
                if assertions:
                    mp = s.get("minProperties")
                    if _jsi_num(mp) and len(keys) < mp:
                        return False
                    mxp = s.get("maxProperties")
                    if _jsi_num(mxp) and len(keys) > mxp:
                        return False
                    req = s.get("required")
                    if isinstance(req, list):
                        for k in req:
                            if k not in inst:
                                return False
                    dr = s.get("dependentRequired")
                    if isinstance(dr, dict):
                        for k, reqs in dr.items():
                            if k in inst:
                                for rk in reqs:
                                    if rk not in inst:
                                        return False
                props = s.get("properties")
                if not isinstance(props, dict):
                    props = {}
                pat_props = s.get("patternProperties")
                if not isinstance(pat_props, dict):
                    pat_props = {}
                for k in keys:
                    matched_lex = False
                    if k in props:
                        if not sub(props[k], inst[k])[0]:
                            return False
                        matched_lex = True
                    for pat, psch in pat_props.items():
                        rex = _jsi_compile_re(pat)
                        if rex is None:
                            continue
                        if rex.search(k) is not None:
                            if not sub(psch, inst[k])[0]:
                                return False
                            matched_lex = True
                    if matched_lex:
                        local["props"].add(k)
                    elif "additionalProperties" in s:
                        if not sub(s["additionalProperties"], inst[k])[0]:
                            return False
                        local["props"].add(k)
                if "propertyNames" in s:
                    pn = s["propertyNames"]
                    for k in keys:
                        if not sub(pn, k)[0]:
                            return False
                dsv = s.get("dependentSchemas")
                if isinstance(dsv, dict):
                    for k, dsch in dsv.items():
                        if k in inst:
                            ok, a = sub(dsch, inst)
                            if not ok:
                                return False
                            merge(a)

            # --- in-place applicators ---
            all_of = s.get("allOf")
            if isinstance(all_of, list):
                for branch in all_of:
                    ok, a = sub(branch, inst)
                    if not ok:
                        return False
                    merge(a)
            any_of = s.get("anyOf")
            if isinstance(any_of, list):
                matched_any = False
                for branch in any_of:
                    ok, a = sub(branch, inst)
                    if ok:
                        matched_any = True
                        merge(a)
                if not matched_any:
                    return False
            one_of = s.get("oneOf")
            if isinstance(one_of, list):
                count = 0
                winner = None
                for branch in one_of:
                    ok, a = sub(branch, inst)
                    if ok:
                        count += 1
                        winner = a
                if count != 1:
                    return False
                if winner is not None:
                    merge(winner)
            if "not" in s:
                if sub(s["not"], inst)[0]:
                    return False
            if "if" in s:
                ok, a = sub(s["if"], inst)
                if ok:
                    merge(a)
                    if "then" in s:
                        ok2, a2 = sub(s["then"], inst)
                        if not ok2:
                            return False
                        merge(a2)
                elif "else" in s:
                    ok2, a2 = sub(s["else"], inst)
                    if not ok2:
                        return False
                    merge(a2)

            # --- unevaluated* (run last, see everything merged above) ---
            if "unevaluatedProperties" in s and isinstance(inst, dict):
                up = s["unevaluatedProperties"]
                for k in inst.keys():
                    if k not in local["props"]:
                        if not sub(up, inst[k])[0]:
                            return False
                        local["props"].add(k)
            if "unevaluatedItems" in s and isinstance(inst, list):
                ui = s["unevaluatedItems"]
                for i in range(len(inst)):
                    if not (local["all"] or i < local["prefix"] or i in local["idxs"]):
                        if not sub(ui, inst[i])[0]:
                            return False
                local["all"] = True

            ann["props"] |= local["props"]
            ann["idxs"] |= local["idxs"]
            if local["prefix"] > ann["prefix"]:
                ann["prefix"] = local["prefix"]
            ann["all"] = ann["all"] or local["all"]
            return True
        finally:
            depth[0] -= 1

    try:
        return vs(root_schema, instance, root_base, [], new_ann())
    except Exception:
        return False

File: crates/json-schema-transformer/src/emit/zod/jsi_validator.ts
type _JsiAnn = { props: Set<string>; prefix: number; all: boolean; idxs: Set<number> };

function _jsiEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (a === null || b === null || typeof a !== 'object' || typeof b !== 'object') return false;
  if (Array.isArray(a) !== Array.isArray(b)) return false;
  if (Array.isArray(a)) {
    const bb = b as unknown[];
    return a.length === bb.length && a.every((v, i) => _jsiEqual(v, bb[i]));
  }
  const ao = a as Record<string, unknown>,
    bo = b as Record<string, unknown>;
  const ak = Object.keys(ao),
    bk = Object.keys(bo);
  if (ak.length !== bk.length) return false;
  return ak.every((k) => Object.prototype.hasOwnProperty.call(bo, k) && _jsiEqual(ao[k], bo[k]));
}

function _jsiValidate(
  rootSchema: unknown,
  remotes: Record<string, unknown>,
  instance: unknown
): boolean {
  const resources = new Map<string, unknown>();
  const anchors = new Map<string, unknown>();
  const dynAnchors = new Map<string, Map<string, unknown>>();
  const baseOf = new Map<object, string>();
  const SKIP = new Set(['const', 'enum', 'default', 'examples']);
  const NAME_MAPS = new Set([
    'properties',
    'patternProperties',
    '$defs',
    'definitions',
    'dependentSchemas',
  ]);

  function stripFrag(u: string): string {
    const i = u.indexOf('#');
    return i === -1 ? u : u.slice(0, i);
  }

  function resolveUri(base: string, ref: string): string {
    if (ref.startsWith('#')) return stripFrag(base) + ref;
    try {
      return new URL(ref, base || undefined).href;
    } catch {
      return ref;
    }
  }

  function indexSchema(node: unknown, base: string): void {
    if (Array.isArray(node)) {
      for (const n of node) indexSchema(n, base);
      return;
    }
    if (node === null || typeof node !== 'object') return;
    const obj = node as Record<string, unknown>;
    let cur = base;
    if (typeof obj.$id === 'string') {
      cur = stripFrag(resolveUri(base, obj.$id));
      resources.set(cur, obj);
    }
    baseOf.set(obj, cur);
    if (typeof obj.$anchor === 'string') anchors.set(cur + '#' + obj.$anchor, obj);
    if (typeof obj.$dynamicAnchor === 'string') {
      let m = dynAnchors.get(cur);
      if (!m) {
        m = new Map();
        dynAnchors.set(cur, m);
      }
      m.set(obj.$dynamicAnchor, obj);
      anchors.set(cur + '#' + obj.$dynamicAnchor, obj);
    }
    for (const [k, v] of Object.entries(obj)) {
      if (SKIP.has(k)) continue;
      if (NAME_MAPS.has(k)) {
        if (v && typeof v === 'object' && !Array.isArray(v)) {
          for (const sub of Object.values(v as Record<string, unknown>)) indexSchema(sub, cur);
        }
        continue;
      }
      indexSchema(v, cur);
    }
  }

  const ROOT_BASE = (() => {
    const id =
      rootSchema && typeof rootSchema === 'object' && !Array.isArray(rootSchema)
        ? (rootSchema as Record<string, unknown>).$id
        : undefined;
    return typeof id === 'string' ? stripFrag(resolveUri('urn:jsi:root', id)) : 'urn:jsi:root';
  })();
  resources.set(ROOT_BASE, rootSchema);
  indexSchema(rootSchema, ROOT_BASE);
  for (const [uri, doc] of Object.entries(remotes)) {
    const u = stripFrag(uri);
    if (!resources.has(u)) resources.set(u, doc);
    indexSchema(doc, u);
  }

  // Honor the dialect's $vocabulary: if the validation vocabulary is absent,
  // assertion keywords (type, const, minimum, required, ...) do not apply.
  let assertions = true;
  {
    const dialect =
      rootSchema && typeof rootSchema === 'object' && !Array.isArray(rootSchema)
        ? (rootSchema as Record<string, unknown>).$schema
        : undefined;
    if (typeof dialect === 'string') {
      const meta = resources.get(stripFrag(dialect));
      if (meta && typeof meta === 'object' && !Array.isArray(meta)) {
        const vocab = (meta as Record<string, unknown>).$vocabulary;
        if (vocab && typeof vocab === 'object' && !Array.isArray(vocab)) {
          assertions = Object.entries(vocab).some(
            ([k, v]) => k.includes('/vocab/validation') && v !== false
          );
        }
      }
    }
  }

  function ptrGet(doc: unknown, ptr: string): unknown {
    if (ptr === '') return doc;
    let cur = doc;
    for (const rawSeg of ptr.split('/').slice(1)) {
      let seg: string;
      try {
        seg = decodeURIComponent(rawSeg);
      } catch {
        seg = rawSeg;
      }
      seg = seg.replace(/~1/g, '/').replace(/~0/g, '~');
      if (Array.isArray(cur)) {
        const i = Number(seg);
        if (!Number.isInteger(i) || i < 0 || i >= cur.length) return undefined;
        cur = cur[i];
      } else if (cur !== null && typeof cur === 'object') {
        if (!Object.prototype.hasOwnProperty.call(cur, seg)) return undefined;
        cur = (cur as Record<string, unknown>)[seg];
      } else {
        return undefined;
      }
    }
    return cur;
  }

  function resolveRef(ref: string, base: string): { schema: unknown; base: string } | null {
    const uri = resolveUri(base, ref);
    const frag = uri.includes('#') ? uri.slice(uri.indexOf('#') + 1) : '';
    const docUri = stripFrag(uri);
    if (frag !== '' && !frag.startsWith('/')) {
      const target = anchors.get(docUri + '#' + frag);
      if (target === undefined) return null;
      return { schema: target, base: baseOf.get(target as object) ?? docUri };
    }
    const doc = resources.get(docUri);
    if (doc === undefined) return null;
    const target = ptrGet(doc, frag);
    if (target === undefined) return null;
    const tBase =
      target !== null && typeof target === 'object' ? baseOf.get(target as object) : undefined;
    return { schema: target, base: tBase ?? docUri };
  }

  let depth = 0;

  function vs(
    schema: unknown,
    inst: unknown,
    base: string,
    dscope: string[],
    ann: _JsiAnn
  ): boolean {
    if (schema === true) return true;
    if (schema === false) return false;
    if (schema === null || typeof schema !== 'object' || Array.isArray(schema)) return true;
    if (++depth > 512) throw new Error('schema recursion limit');
    try {
      const s = schema as Record<string, unknown>;
      let myBase = baseOf.get(schema as object) ?? base;
      let scope = dscope;
      if (scope.length === 0 || scope[scope.length - 1] !== myBase) scope = [...scope, myBase];

      const local: _JsiAnn = { props: new Set(), prefix: 0, all: false, idxs: new Set() };
      const merge = (child: _JsiAnn) => {
        for (const p of child.props) local.props.add(p);
        for (const i of child.idxs) local.idxs.add(i);
        local.prefix = Math.max(local.prefix, child.prefix);
        local.all = local.all || child.all;
      };
      const sub = (sch: unknown, i: unknown): { ok: boolean; ann: _JsiAnn } => {
        const a: _JsiAnn = { props: new Set(), prefix: 0, all: false, idxs: new Set() };
        return { ok: vs(sch, i, myBase, scope, a), ann: a };
      };

      // --- $ref / $dynamicRef (in-place applicators) ---
      if (typeof s.$ref === 'string') {
        const r = resolveRef(s.$ref, myBase);
        if (!r) return false;
        const a: _JsiAnn = { props: new Set(), prefix: 0, all: false, idxs: new Set() };
        if (!vs(r.schema, inst, r.base, scope, a)) return false;
        merge(a);
      }
      if (typeof s.$dynamicRef === 'string') {
        const r = resolveRef(s.$dynamicRef, myBase);
        if (!r) return false;
        let target = r.schema;
        let tBase = r.base;
        const frag = s.$dynamicRef.includes('#')
          ? s.$dynamicRef.slice(s.$dynamicRef.indexOf('#') + 1)
          : '';
        const isPlain = frag !== '' && !frag.startsWith('/');
        const tObj =
          target !== null && typeof target === 'object'
            ? (target as Record<string, unknown>)
            : null;
        if (isPlain && tObj && tObj.$dynamicAnchor === frag) {
          for (const res of scope) {
            const m = dynAnchors.get(res);
            if (m && m.has(frag)) {
              target = m.get(frag);
              tBase = baseOf.get(target as object) ?? res;
              break;
            }
          }
        }
        const a: _JsiAnn = { props: new Set(), prefix: 0, all: false, idxs: new Set() };
        if (!vs(target, inst, tBase, scope, a)) return false;
        merge(a);
      }

      // --- type ---
      if (assertions && s.type !== undefined) {
        const types = Array.isArray(s.type) ? (s.type as string[]) : [s.type as string];
        const ok = types.some((t) => {
          switch (t) {
            case 'null':
              return inst === null;
            case 'boolean':
              return typeof inst === 'boolean';
            case 'string':
              return typeof inst === 'string';
            case 'number':
              return typeof inst === 'number';
            case 'integer':
              return typeof inst === 'number' && Number.isInteger(inst);
            case 'array':
              return Array.isArray(inst);
            case 'object':
              return inst !== null && typeof inst === 'object' && !Array.isArray(inst);
            default:
              return false;
          }
        });
        if (!ok) return false;
      }

      // --- const / enum ---
      if (assertions && s.const !== undefined && !_jsiEqual(inst, s.const)) return false;
      if (assertions && Array.isArray(s.enum) && !s.enum.some((e) => _jsiEqual(inst, e)))
        return false;

      // --- numeric ---
      if (assertions && typeof inst === 'number') {
        if (typeof s.multipleOf === 'number') {
          const q = inst / (s.multipleOf as number);
          if (!Number.isFinite(q) || q !== Math.round(q)) return false;
        }
        if (typeof s.minimum === 'number' && inst < s.minimum) return false;
        if (typeof s.maximum === 'number' && inst > s.maximum) return false;
        if (typeof s.exclusiveMinimum === 'number' && inst <= s.exclusiveMinimum) return false;
        if (typeof s.exclusiveMaximum === 'number' && inst >= s.exclusiveMaximum) return false;
      }

      // --- string ---
      if (assertions && typeof inst === 'string') {
        const cp = () => Array.from(inst).length;
        if (typeof s.minLength === 'number' && cp() < s.minLength) return false;
        if (typeof s.maxLength === 'number' && cp() > s.maxLength) return false;
        if (typeof s.pattern === 'string') {
          let re: RegExp;
          try {
            re = new RegExp(s.pattern, 'u');
          } catch {
            try {
              re = new RegExp(s.pattern);
            } catch {
              re = /(?:)/;
            }
          }
          if (!re.test(inst)) return false;
        }
      }

      // --- array ---
      if (Array.isArray(inst)) {
        if (assertions) {
          if (typeof s.minItems === 'number' && inst.length < s.minItems) return false;
          if (typeof s.maxItems === 'number' && inst.length > s.maxItems) return false;
          if (s.uniqueItems === true) {
            for (let i = 0; i < inst.length; i++)
              for (let j = i + 1; j < inst.length; j++)
                if (_jsiEqual(inst[i], inst[j])) return false;
          }
        }
        const prefix = Array.isArray(s.prefixItems) ? (s.prefixItems as unknown[]) : [];
        for (let i = 0; i < Math.min(prefix.length, inst.length); i++) {
          if (!sub(prefix[i], inst[i]).ok) return false;
        }
        if (prefix.length > 0)
          local.prefix = Math.max(local.prefix, Math.min(prefix.length, inst.length));
        if (s.items !== undefined) {
          for (let i = prefix.length; i < inst.length; i++) {
            if (!sub(s.items, inst[i]).ok) return false;
          }
          local.all = true;
        }
        if (s.contains !== undefined) {
          const matched: number[] = [];
          for (let i = 0; i < inst.length; i++) {
            if (sub(s.contains, inst[i]).ok) matched.push(i);
          }
          const minC = typeof s.minContains === 'number' ? s.minContains : 1;
          const maxC = typeof s.maxContains === 'number' ? s.maxContains : Infinity;
          if (matched.length < minC || matched.length > maxC) return false;
          for (const i of matched) local.idxs.add(i);
        }
      }

      // --- object ---
      if (inst !== null && typeof inst === 'object' && !Array.isArray(inst)) {
        const io = inst as Record<string, unknown>;
        const keys = Object.keys(io);
        if (assertions) {
          if (typeof s.minProperties === 'number' && keys.length < s.minProperties) return false;
          if (typeof s.maxProperties === 'number' && keys.length > s.maxProperties) return false;
          if (Array.isArray(s.required)) {
            for (const k of s.required as string[]) {
              if (!Object.prototype.hasOwnProperty.call(io, k)) return false;
            }
          }
          if (s.dependentRequired && typeof s.dependentRequired === 'object') {
            for (const [k, reqs] of Object.entries(
              s.dependentRequired as Record<string, string[]>
            )) {
              if (Object.prototype.hasOwnProperty.call(io, k)) {
                for (const r of reqs)
                  if (!Object.prototype.hasOwnProperty.call(io, r)) return false;
              }
            }
          }
        }
        const props =
          s.properties && typeof s.properties === 'object'
            ? (s.properties as Record<string, unknown>)
            : {};
        const patProps =
          s.patternProperties && typeof s.patternProperties === 'object'
            ? (s.patternProperties as Record<string, unknown>)
            : {};
        for (const k of keys) {
          let matchedLexical = false;
          if (Object.prototype.hasOwnProperty.call(props, k)) {
            if (!sub(props[k], io[k]).ok) return false;
            matchedLexical = true;
          }
          for (const [pat, psch] of Object.entries(patProps)) {
            let re: RegExp;
            try {
              re = new RegExp(pat, 'u');
            } catch {
              try {
                re = new RegExp(pat);
              } catch {
                continue;
              }
            }
            if (re.test(k)) {
              if (!sub(psch, io[k]).ok) return false;
              matchedLexical = true;
            }
          }
          if (matchedLexical) local.props.add(k);
          if (!matchedLexical && s.additionalProperties !== undefined) {
            if (!sub(s.additionalProperties, io[k]).ok) return false;
            local.props.add(k);
          }
        }
        if (s.propertyNames !== undefined) {
          for (const k of keys) {
            if (!sub(s.propertyNames, k).ok) return false;
          }
        }
        if (s.dependentSchemas && typeof s.dependentSchemas === 'object') {
          for (const [k, dsch] of Object.entries(s.dependentSchemas as Record<string, unknown>)) {
            if (Object.prototype.hasOwnProperty.call(io, k)) {
              const r = sub(dsch, inst);
              if (!r.ok) return false;
              merge(r.ann);
            }
          }
        }
      }

      // --- in-place applicators ---
      if (Array.isArray(s.allOf)) {
        for (const branch of s.allOf as unknown[]) {
          const r = sub(branch, inst);
          if (!r.ok) return false;
          merge(r.ann);
        }
      }
      if (Array.isArray(s.anyOf)) {
        let any = false;
        for (const branch of s.anyOf as unknown[]) {
          const r = sub(branch, inst);
          if (r.ok) {
            any = true;
            merge(r.ann);
          }
        }
        if (!any) return false;
      }
      if (Array.isArray(s.oneOf)) {
        let count = 0;
        let winner: _JsiAnn | null = null;
        for (const branch of s.oneOf as unknown[]) {
          const r = sub(branch, inst);
          if (r.ok) {
            count++;
            winner = r.ann;
          }
        }
        if (count !== 1) return false;
        if (winner) merge(winner);
      }
      if (s.not !== undefined) {
        if (sub(s.not, inst).ok) return false;
      }
      if (s.if !== undefined) {
        const ifR = sub(s.if, inst);
        if (ifR.ok) {
          merge(ifR.ann);
          if (s.then !== undefined) {
            const r = sub(s.then, inst);
            if (!r.ok) return false;
            merge(r.ann);
          }
        } else if (s.else !== undefined) {
          const r = sub(s.else, inst);
          if (!r.ok) return false;
          merge(r.ann);
        }
      }

      // --- unevaluated* (run last, see everything merged above) ---
      if (
        s.unevaluatedProperties !== undefined &&
        inst !== null &&
        typeof inst === 'object' &&
        !Array.isArray(inst)
      ) {
        const io = inst as Record<string, unknown>;
        for (const k of Object.keys(io)) {
          if (!local.props.has(k)) {
            if (!sub(s.unevaluatedProperties, io[k]).ok) return false;
            local.props.add(k);
          }
        }
      }
      if (s.unevaluatedItems !== undefined && Array.isArray(inst)) {
        const covered = (i: number) => local.all || i < local.prefix || local.idxs.has(i);
        for (let i = 0; i < inst.length; i++) {
          if (!covered(i)) {
            if (!sub(s.unevaluatedItems, inst[i]).ok) return false;
          }
        }
        local.all = true;
      }

      for (const p of local.props) ann.props.add(p);
      for (const i of local.idxs) ann.idxs.add(i);
      ann.prefix = Math.max(ann.prefix, local.prefix);
      ann.all = ann.all || local.all;
      return true;
    } finally {
      depth--;
    }
  }

  try {
    const ann: _JsiAnn = { props: new Set(), prefix: 0, all: false, idxs: new Set() };
    return vs(rootSchema, instance, ROOT_BASE, [], ann);
  } catch {
    return false;
  }
}
